tests/trainer/test_fsdp.py

```diff
@@ -272,3 +272,59 @@ def test_fsdp_act_ckpt_offload(
             assert isinstance(trainer.state.model.fc1._fsdp_wrapped_module, OffloadWrapper)
         else:
             assert not isinstance(trainer.state.model.fc1._fsdp_wrapped_module, CheckpointWrapper)
+
+
+class SimpleMLPForTestingOOM(ComposerModel):
+
+    def __init__(self, num_features: int = 128, device: str = 'cuda'):
+        super().__init__()
+        self.device = device
+        self.fc1 = torch.nn.Linear(num_features, num_features, device=device, bias=False)
+        self.fc2 = torch.nn.Linear(num_features, num_features, device=device, bias=False)
+        self.fc3 = torch.nn.Linear(num_features, num_features, device=device, bias=False)
+        self.rank = dist.get_global_rank()
+
+        def oom_hook(*args):
+            raise RuntimeError('CUDA out of memory.')
+
+        self.fc2.register_full_backward_hook(oom_hook)
+
+    def forward(self, x):
+        x = self.fc1(x)
+        x = self.fc2(x)
+        x = self.fc3(x)
+        return x
+
+    def loss(self, outputs, batch):
+        return torch.sum(outputs)
+
+
+@pytest.mark.gpu
+@world_size(2)
+def test_fsdp_reshard_after_oom(world_size: int):
+    model = SimpleMLPForTestingOOM()
+
+    trainer = Trainer(
+        model=model,
+        fsdp_config={},
+        max_duration='3ba',
+        dist_timeout=20,
+    )
+    fsdp_model = trainer.state.model
+
+    x = torch.rand([2, 128])
+    output = fsdp_model(x)
+    with pytest.raises(Exception):
+        # Backward triggers the fake OOM exception,
+        # which prevents fsdp reshard and cleanup
+        torch.sum(output).backward()
+
+    fc2_flat_param = fsdp_model.fc2._flat_param
+
+    # without cleanup, model.fc2.flat_params is still in unshard state
+    # the full param is not freed
+    assert fc2_flat_param.data_ptr() != fc2_flat_param._local_shard.data_ptr()
```

Comment:
> if the param is not freed, the data ptr of the flat param will be different than the data ptr of the local shard

basically yes, looking at the flat_param.reshard(), it has [two high level steps](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_flat_param.py#L1705-L1707): 
1. switch the flat_param.data_ptr() from `_full_param_padded` to `_local_shard`
2. free the storage of `full_params_padded`

> _full_param_padded part, is that just empty if the full param has been freed?
I think yes. 
I tried adding `assert fc2_flat_param._full_param_padded.numel() == 0` after the reshard and cleanup, but it failed, because some internal tensor check. i guess the storage is not freed in a normal way. After freeing the storage, it's not a valid tensor anymore, any tensor api call will cause a internal tensor check error.

Reply:
Can we test that second piece via whatever method you've found reliable to tell? like a with pytest.raises on some tensor api call on it. we're relying on internal apis and undocumented behavior here, so would be good to have unit test as thorough as possible.